Add ability to copy theme mods from active theme to reader theme in AMP Customizer

westonruter opened the pull request:
## Summary

Given that the active theme has a Custom Logo set and has assigned a nav menu to the Primary and Social nav menu locations, when opening the AMP Customizer when the Reader theme does _not_ have the same values assigned to its settings, a new “Active Theme's Settings” section appears:

![image](https://user-images.githubusercontent.com/134745/88892123-2a99b600-d1f9-11ea-93eb-a2cc04bb42bd.png)

Expanding the section reveals the controls that have settings which can be imported from the active theme, grouped by the sections that the controls appear in:

![image](https://user-images.githubusercontent.com/134745/88892343-74829c00-d1f9-11ea-8e72-5c4df9873cc3.png)

All settings are checked by default, so that if the Import button is clicked without ever having expanded the section, then all the settings will be imported.

If only the checkbox next to the Site Logo control is checked, then only the setting(s) for that control are imported and then the item is removed from the list:

![image](https://user-images.githubusercontent.com/134745/88892441-a09e1d00-d1f9-11ea-8949-1e466d5ae59e.png)

If all the checkboxes were checked when the Import button was clicked, then all settings are imported and the entire section is removed:

![image](https://user-images.githubusercontent.com/134745/88892508-bca1be80-d1f9-11ea-938c-bd2cb29e8e0d.png)

Fixes #5039
Fixes #5137

## Todo

- [ ] Clean up the classic Customizer JS logic as much as possible.
- [ ] Tests.

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

westonruter committed:
Improve placement of register method

westonruter committed:
Add naive method of copying theme mods to reader theme

westonruter committed:
Remove naive theme mod copying theme mods from active theme to reader theme

westonruter committed:
Add initial plumbing to import theme mods from active theme to reader theme

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/reader-theme-mod-sync

* 'develop' of github.com:ampproject/amp-wp: (393 commits)
  E2E test: scroll to item before clicking it
  Fix e2e test expecting wizard done screen not to have close button for reader mode
  Use constant for 'reader'
  Fix failing 'return to previous page' e2e test
  Wizard: show close button on last page when reader mode was selected
  Reduce delay for showing iframe on Done screen
  Fix mobile redirection e2e test
  Fix failing Jest test
  Onboarding wizard: change "Summary" to "Review"
  Supply newly-missing ImageObject type for publisher logo in Schema.org metadata
  Fix confusing initialOpen logic for drawers
  Minor suggestions from code review
  Remove unused const
  Remove unused ref
  Use passive event listeners
  Small concision improvement
  Remove unneeeded comma
  Install and integrate classnames library
  Finish sentence with period
  Remove references to removed intersection observer
  ...

westonruter committed:
Fix populating background image and header image controls

westonruter committed:
Fix importing nav menus

westonruter committed:
Fix preparing Customizer setting PHP values for JS

westonruter committed:
Fix broken styling of panel preview-notice

westonruter committed:
Add list of settings to import

westonruter committed:
Move more translation strings to JS

westonruter committed:
Allow fine-grained importing of settings

westonruter committed:
Remove unused importSettings function

westonruter committed:
Fix stylelint issue

westonruter committed:
Add support for importing Custom CSS from active theme

westonruter committed:
Ensure JS-preparation filters apply to exported settings

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/reader-theme-mod-sync

* 'develop' of github.com:ampproject/amp-wp: (59 commits)
  Fix import of reader constant
  Update active page index after render phase
  Fix test_print_edit_form_validation_status
  Update index.js
  Update assets/src/onboarding-wizard/pages/welcome/index.js
  Update assets/src/onboarding-wizard/pages/template-mode/get-selection-details.js
  Update assets/src/settings-page/template-modes.js
  Update assets/src/onboarding-wizard/pages/save/index.js
  Update PluginActivationNotice.php
  Update src/Admin/PluginActivationNotice.php
  Update src/Admin/PluginActivationNotice.php
  Update includes/validation/class-amp-validation-manager.php
  Update includes/validation/class-amp-validation-error-taxonomy.php
  Update includes/validation/class-amp-validation-error-taxonomy.php
  Update includes/validation/class-amp-validation-error-taxonomy.php
  Update includes/validation/class-amp-validated-url-post-type.php
  Update includes/validation/class-amp-validated-url-post-type.php
  Update includes/validation/class-amp-validated-url-post-type.php
  Fix erroneous file comment error
  Remove unused 'use'
  ...

westonruter committed:
Add 'AMP Settings' button to Reader theme lightbox

westonruter committed:
Use 'Customize' instead of 'Customize AMP' for button text in Reader theme lightbox

westonruter committed:
Let finish button in Reader mode wizard say 'Customize Theme'

westonruter committed:
Rename 'Visit Customizer' link to 'Customize Theme' on admin screen

jwold commented:
This looks good @westonruter for a start!

westonruter committed:
Improve margins for dl

westonruter committed:
Use primary as opposed to active when talking about theme for settings import

westonruter committed:
Make Primary Theme Settings bold and cursor:pointer

amedina reviewed (approved):
Great capability which goes a long way in helping publishers to achieve brand affinity. 

One variation I would like to discuss (after release) is to keep the UI component present but grayed out when settings are the same. The rationale is that users will remain aware of the capability, and since it can be collapsed, keeping the component around would not be intrusive.

westonruter committed:
Keep track of timestamps for when theme mod settings were modified in active and reader theme

westonruter committed:
Add descriptive text to Primary Theme Settings

pierlon commented on assets/css/src/amp-customizer.css:
```diff
@@ -1,5 +1,54 @@
-#customize-controls .customize-info .preview-notice {
+#customize-controls #customize-info .preview-notice {
```
```suggestion
#customize-info .preview-notice {
```

westonruter replied:
Unfortunately this won't work. It needs to be this specific to override this selector in core: `#customize-controls .customize-info .preview-notice`.

See https://github.com/WordPress/wordpress-develop/blob/6d3207d564c6eb19de5d0f0532b30903a97a533a/src/wp-admin/css/customize-controls.css#L429-L432

pierlon commented on assets/src/customizer/amp-customize-controls.js:
```diff
@@ -1,5 +1,15 @@
 /* global jQuery */
 
+/**
```
Not an expert on the Customizer API so I'm deferring to your expertise. The code seems sound and is functional so I'm giving it a :+1:.

pierlon commented on assets/src/settings-page/welcome.js:
```diff
@@ -92,7 +92,7 @@ export function Welcome() {
 						customizerLink && (
 							<a className="components-button is-link" href={ customizerLink } rel="noreferrer">
-								{ __( 'Visit Customizer', 'amp' ) }
+								{ __( 'Customize Theme', 'amp' ) }
```
I'm wondering if this could be confusing for the user. I think specifying what theme they're customizing would prevent any potential confusion.

```suggestion
								{ __( 'Customize Reader Theme', 'amp' ) }
```

pierlon commented on assets/css/src/amp-customizer.css:
```diff
@@ -1,5 +1,54 @@
+
+#customize-controls .control-section-amp_active_theme_settings_import > .accordion-section-title summary {
+	cursor: pointer;
+	display: inline-block;
```
The summary marker is not being shown in Firefox:

![image](https://user-images.githubusercontent.com/16200219/88984679-c4c33200-d2bd-11ea-9658-9afab8253f36.png)

 It seems if the `display` rule for the `summary` element does not match its parent's, the marker is not shown.

```suggestion
	display: list-item;
```

westonruter committed:
Make clear Customize link is for Reader theme

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Fix display of summary marker in Firefox

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter commented:
@pierlon In regards to 2b029f5, one thing that makes me nervous is the placement of the Import button with the `summary`. I had to limit the width to 75%:

![image](https://user-images.githubusercontent.com/134745/88986043-b86fb280-d286-11ea-8036-2fe81a9d8678.png)

However, if the Import button is longer, then it's not going to work well:

![image](https://user-images.githubusercontent.com/134745/88986195-287e3880-d287-11ea-8e51-bc682ff877c1.png)

I suppose I should just opt for fixed positioning for the Import button.

westonruter commented:
Or no, the absolute positioning makes it worse since then the can overlap the text. What we need is to disable the focus outline I guess.

westonruter committed:
Use focus color instead of outline

westonruter commented:
I've addressed this by using a focus color instead of a focus outline.

Unfocused:

![image](https://user-images.githubusercontent.com/134745/88987409-b3146700-d28a-11ea-8e4c-c4810caa4c62.png)

Focused:

![image](https://user-images.githubusercontent.com/134745/88987433-bf002900-d28a-11ea-8e0f-6dd74c7c444e.png)

This somewhat makes sense because the other sections use blue for the focus styling:

![image](https://user-images.githubusercontent.com/134745/88987470-d8a17080-d28a-11ea-98dd-d1775d053218.png)

pierlon committed:
Use flex layout for settings import section

pierlon commented:
@googlebot I consent.

westonruter committed:
Fix static analysis of get_active_theme_import_settings

westonruter committed:
Improve jsdoc

westonruter commented on includes/admin/class-amp-template-customizer.php:
```diff
@@ -332,12 +350,181 @@ public function add_customizer_scripts() {
 			amp_get_asset_url( 'css/amp-customizer.css' ),
 			[],
-			AMP__VERSION
+			$version
```
@pierlon This change isn't right as the `$version` here is for the script not the style.

```suggestion
			AMP__VERSION
```

pierlon replied:
Yea that was not intentional. The browser was not loading the changes I made to that CSS file so I used `$version` to bust the cache.

westonruter committed:
Revert "Use flex layout for settings import section"

This reverts commit 0d0fa02599f604a1b95c6a0bb1b72ba4ff7ce582.

westonruter committed:
Add test for initializing non-AMP Customizer when Reader theme selected

westonruter committed:
Add test for store_modified_theme_mod_setting_timestamps and fix setting type check

westonruter committed:
Account for WP_Customize_Filter_Setting

westonruter committed:
Add test for get_active_theme_import_settings

westonruter committed:
Add test for AMP_Template_Customizer::render_setting_import_section_template

westonruter committed:
Fix WP 4.9 compatibility when wp_set_script_translations() does not exist

westonruter committed:
Fix importing setting for Display Site Title and Tagline

Fixes #5137

pierlon commented:
With a child of Twenty Twenty active and Twenty Twenty as Reader theme, the "Header & Footer Background Color" setting label is incorrectly escaped and the ~"Accent hue"~ "Primary Color" label is not displayed at all:

![image](https://user-images.githubusercontent.com/16200219/89003453-55663600-d2ef-11ea-8d74-9ddb2c1931c6.png)

pierlon commented:
With the same scenario described above, the "Overlay Opacity" setting (found in Cover Template section) is not listed to be imported.

westonruter committed:
Break up ready method into separate methods

westonruter committed:
Fix displaying of escaped HTML in checkbox labels

westonruter committed:
Fix handling import of background settings

westonruter committed:
Refactor handling of checkbox controls with synced elements

westonruter committed:
Refine jsdoc and eslint issues

westonruter committed:
Omit background image settings from import if no background_image set

westonruter committed:
Only include controls which have labels

westonruter committed:
Fix importing accent_hue color from Twenty Twenty child theme

westonruter commented:
@pierlon Please check again. I believe I've addressed the issues you raised.

![image](https://user-images.githubusercontent.com/134745/89069477-10122a80-d328-11ea-96c4-ab960a15b395.png)

westonruter committed:
Add test for excluding background_image-related control settings if background_image empty

westonruter committed:
Fix including controls that have related settings that differ

pierlon commented on assets/src/customizer/amp-customize-controls.js:
```diff
@@ -84,6 +98,419 @@ window.ampCustomizeControls = ( function( api, $ ) {
+	];
+
+	/**
+	 * Mapping of control ID to the setting value which indicates the checkbox should *unchecked*.
```
```suggestion
	 * Mapping of control ID to the setting value which indicates the checkbox should be unchecked.
```

pierlon commented on assets/src/customizer/amp-customize-controls.js:
```diff
@@ -84,6 +98,419 @@ window.ampCustomizeControls = ( function( api, $ ) {
+	 *
+	 * Because of some "juggling" in WordPress core, programmatically updating the value (probably related to double
+	 * data binding of Element and Setting values), updating this control by directly manipulating the Element instance
+	 * as opposed to the underlying Setting that it syncs with.
```
```suggestion
	 * Because of some "juggling" in WordPress core, programmatically updating the value (probably related to double
	 * data binding of Element and Setting values) is not feasible. It is instead updated by directly manipulating the Element instance
	 * as opposed to the underlying Setting that it syncs with.
```

westonruter committed:
Improve JS comments

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>
